Housnav-21: Walkthrough Card Component

wardm89 opened the pull request:
<!-- You may remove any sections that are not applicable -->

[HOUSNAV-21](https://hous-bssb.atlassian.net/browse/HOUSNAV-21)

## Overview & Purpose
<!-- Please describe the purpose of your changes. Please also include relevant motivation and context. -->
Add a reusable walkthrough card component to be used through the app. (Home page and end result)

## Summary of Changes
<!-- Please include a HIGH LEVEL overview of the code changes. (Added xyz fields to table, modified function to handle x case instead of y case, etc. )  -->
- Adds `walkthoughCard` component
- Updates the `9.9.9.` walkthrough to add titles and descriptions for the cards
    - Updates the `WalkthroughInfo` accordingly 

## Side Effects
<!-- Any possible side effects? Does this change affect features that are not linked to the direct purpose? -->
- Adjusted `useWalkthroughData` error handling to make it more understandable

## Testing
<!-- Please include steps to test out the changes/fixes if applicable OR context notes to help get env in a state to test-->
- Go to home screen and new card should appear
- It should navigate to the 9.9.9 walkthrough when clicked

https://github.com/bcgov/House-Innovations-HOUSNAV/assets/7059124/52b6de35-7346-4a12-9226-49f6de218863


![Screenshot 2024-06-13 at 9 59 12 AM](https://github.com/bcgov/House-Innovations-HOUSNAV/assets/7059124/88f65fd9-ea2c-4345-ba8b-aff4668c7177)

![Screenshot 2024-06-13 at 9 59 30 AM](https://github.com/bcgov/House-Innovations-HOUSNAV/assets/7059124/e110675c-b14c-4fc7-a095-7ab313b874e1)

wardm89 committed:
First pass adding walkthrough card

css changes needed

wardm89 committed:
css walkthrough card adjusments

Updated walkthrough iteration

remove export

css updates

wardm89 committed:
Remove duplicate card logic

wardm89 committed:
Adds walkthrough card test

jbook24 commented:
Could we apply an ease-in-out transition on hover? maybe 300ms?
https://stackoverflow.com/questions/16163922/css3-transition-ease-in-and-out-box-shadow

jbook24 commented on apps/web/app/page.tsx:
```diff
@@ -1,15 +1,24 @@
 
 export default function Page(): JSX.Element {
   return (
     <div className="u-container">
```
Should this div be a section? 
```
<section> 
   <div>
```

wardm89 replied:
The landing page html will be done with https://hous-bssb.atlassian.net/jira/software/c/projects/HOUSNAV/boards/3/backlog?selectedIssue=HOUSNAV-30

jbook24 commented on apps/web/app/page.tsx:
```diff
@@ -1,15 +1,24 @@
-      <p>
-        <Link href={`${URL_WALKTHROUGH_HREF}/9.9.9/`}>Walkthrough 9.9.9</Link>
-      </p>
+      <div>
```
Same here

wardm89 committed:
adds transition to walkthrough card hover state

wardm89 commented:
> Could we apply an ease-in-out transition on hover? maybe 300ms? https://stackoverflow.com/questions/16163922/css3-transition-ease-in-and-out-box-shadow

Added a transition to the shadow and underline

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,47 @@
+  if (!canDisplayCard) return null;
+
+  return (
+    <Button
```
I think semantically this should be an `<a>` tag and not a button as it's sole purpose is to navigate via an href.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,47 @@
+      onPress={handleClick}
+      aria-label={`${id}-walkthrough-card`}
+    >
+      <header className="ui-WalkthroughCard--Title">{data.title}</header>
```
Also I'm not a semantic expert, but I **_think_** the structure we want here is something like:

```html
<article>
  <header>
    <h2/h3> title -- depends on where it's being used
    <p> subtitle
  </header>
  <p> description
</article>
```

nicholaslyon-dt commented on packages/data/walkthroughs/9.9.9.json:
```diff
@@ -33,27 +35,11 @@
-        "P23",
-        "P24"
-      ]
+      "sectionQuestions": ["P18", "P19", "P20", "P21", "P22", "P23", "P24"]
```
Did vscode auto do this for you or was this manual. My IDE formats it one on each line.

wardm89 replied:
Yes mine did that automatically when I saved.

nicholaslyon-dt replied:
Alright. I updated my IDE to "wrap if long" hopefully it'll leave it as is from now on.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.css:
```diff
@@ -0,0 +1,46 @@
+.ui-WalkthroughCard--CardContainer {
+  display: flex;
+  border-radius: var(--layout-border-radius-xlarge);
+  box-shadow: var(--surface-shadow-none), var(--surface-shadow-none);
```
I'm seeing `box-shadow: var(--surface-shadow-medium);` in the designs. Also the definitions in `variables.css` already have both parts with a comma.

![Screenshot 2024-06-17 at 11 15 07 AM](https://github.com/bcgov/House-Innovations-HOUSNAV/assets/41478918/5f0229d0-d938-470a-9d71-2c3e83714645)

wardm89 replied:
Yep, that's what it transitions to on hover

nicholaslyon-dt replied:
I'm seeing a transition from medium -> large on hover. Are you seeing something different?

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.css:
```diff
@@ -0,0 +1,46 @@
+  text-underline-offset: var(--walkthrough-card-title-text-underline-offset);
+}
+
+.ui-WalkthroughCard--CardContainer:hover {
```
Can we use [data-hovered] to match the paradigms that react aria uses? Take a look at how button.css/link.css are handling it. We might want to also consider something for the [data-focus-visible] state as well.

wardm89 replied:
Since the card isn't a specific aria component like the button or link then no we can't use that...

nicholaslyon-dt replied:
Ahh, previously it was using the Button component so we could have.

wardm89 committed:
Updating router mock

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.css:
```diff
@@ -0,0 +1,46 @@
+
+.ui-WalkthroughCard--CardContainer:hover {
+  cursor: pointer;
+  box-shadow: var(--surface-shadow-medium), var(--surface-shadow-medium);
```
Same as above, I'm seeing large for this state.

![Screenshot 2024-06-17 at 11 15 19 AM](https://github.com/bcgov/House-Innovations-HOUSNAV/assets/41478918/a28307ad-96c2-4427-b60d-93a13d4baafc)

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.css:
```diff
@@ -0,0 +1,46 @@
+
+.ui-WalkthroughCard--Title {
+  color: var(--support-border-color-info);
+  font: var(--typography-bold-h5);
```
This looks like it needs to be h4 for the homepage cards and h5 for the result page cards based on the design.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.css:
```diff
@@ -0,0 +1,46 @@
+  background-color: var(--surface-color-background-white);
+  max-width: var(--walkthrough-card-max-width);
+  flex-direction: column;
+  font-size: 14px;
```
Do we need this font declaration here? I think all the content inside has it's own font declarations right? Same with the color line below this.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.css:
```diff
@@ -0,0 +1,46 @@
+  font-size: 14px;
+  color: var(--support-border-color-info);
+  padding: var(--layout-padding-large);
+  margin: var(--layout-margin-medium);
```
I would assume in our use cases for these I'd want the container to some gap spacing to the cards. This makes a minimum spacing on these that can't be reduced by the parent easily.

nicholaslyon-dt replied:
I'm not seeing this changed. Thoughts on removing it?

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,47 @@
+    if (href) router.push(href);
+  };
+
+  const canDisplayCard =
```
I don't think we need this check based on all the types and properties being required. We'd get an error in useWalkthroughData.ts if any imported JSON didn't have the appropriate WalkthroughInfo.

wardm89 replied:
True

nicholaslyon-dt replied:
Any reason I'm missing why we still have it then?

nicholaslyon-dt commented on apps/web/app/page.tsx:
```diff
@@ -1,16 +1,24 @@
+      <div>
+        {Object.entries(WalkthroughJSONData).map(([id, { info }]) => {
+          return (
+            <WalkthroughCard
```
These need a key property (ignore if you'd planned to cover that when building the actual page)

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,47 @@
+
+import { Button } from "react-aria-components";
+
+import { WalkthroughInfo } from "../../../data/src/useWalkthroughData";
```
This needs to import from the package.

`@repo/data/useWalkthroughData`

nicholaslyon-dt commented on apps/web/app/page.tsx:
```diff
@@ -1,16 +1,24 @@
 import { URL_WALKTHROUGH_HREF } from "@repo/constants/src/urls";
-import ModalSide from "@repo/ui/modal-side";
+import WalkthroughCard from "@repo/ui/walkthrough-card";
+import { WalkthroughJSONData } from "../../../packages/data/src/useWalkthroughData";
```
Needs to use the `@repo/data..` import

wardm89 replied:
Nice I didn't know @repo was configured to support that. This is why I mentioned adding absolute imports! haha

nicholaslyon-dt commented on packages/data/src/useWalkthroughData.ts:
```diff
@@ -75,7 +75,7 @@ export interface QuestionMultipleChoiceData extends QuestionBaseData {
 export const WalkthroughItemTypeMultiChoiceMultiple = "multiChoiceMultiple";
 export const isWalkthroughItemTypeMultiChoiceMultiple = (
-  walkthroughItemType: string,
+  walkthroughItemType: string
```
My prettier puts this back in...

Can you run `npm run format` from the root? That _**should**_ format everything appropriately.

nicholaslyon-dt commented on packages/data/src/useWalkthroughData.ts:
```diff
@@ -122,15 +126,21 @@ export interface WalkthroughJSONType {
+export default function useWalkthroughData({
+  id,
+}: UseWalkthroughDataProps): WalkthroughJSONType {
+  if (!id) {
```
Since id is a required prop by our types, do we need this?

wardm89 replied:
I think that's fine and the other comment about not needing the check as long as we are on the same page with the assumption that the walkthrough `json` is in fact correct and should bomb the application (in lower envs) if we somehow mess it up and pass incorrect data.

nicholaslyon-dt reviewed (changes requested):
Sorry for all the comments! It looks great overall.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,45 @@
+  if (!canDisplayCard) return null;
+
+  return (
+    <a
```
Sorry, I should have said this different the first time. This should use either the react-aria-components Link or the next/link to take advantage of client side routing. This way we can just add an href to it and it'll also be accessible. We don't need to handle any clicking or use the router here.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,45 @@
+export interface WalkthroughCardProps {
+  id: string;
+  data: WalkthroughInfo;
+  href?: string;
```
I think this should be required based on our use cases. What would your thoughts be about making this be a property called `walkthroughId` and moving the pathing part with `URL_WALKTHROUGH_HREF` directly into the component. In theory the 2 use cases for this always point to a walkthrough. We could also create a helper function in `url.ts` to replace `URL_WALKTHROUGH_HREF` like we've done with some of the testids like so:

```TypeScript
export const GET_WALKTHROUGH_HREF = (id: string) => `/walkthrough/${id}/`;
```

wardm89 replied:
I like the idea of moving it to the card actually. The helper function I don't think is necessary since we already have a constant for the base path for walkthroughs and can just do 
`href={`${URL_WALKTHROUGH_HREF}/${walkthroughId}`}`

nicholaslyon-dt reviewed (changes requested):
Updated comments and callouts!

wardm89 committed:
PR comment updates

PR comment updates(2)

Update walkthrough card css and props

Update walkthrough card css shadow

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,39 @@
+export default function WalkthroughCard({
+  id,
+  data,
+  walkthroughId,
```
Since we're extending the props, we should probably catch them here with `...props` and pass them along to ReactAriaLink.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.css:
```diff
@@ -0,0 +1,45 @@
+.ui-WalkthroughCard--CardContainer[data-hovered] {
+  cursor: pointer;
+  box-shadow: var(--surface-shadow-large), var(--surface-shadow-large);
+  border: 1px solid var(--surface-color-border-default);
```
Good call on the border on hover. Completely missed that before. Can we add it as a transparent border by default and then just change the border color when hovering? I think the animation looks better that way. Also update the transition declaration to include boder-color.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.css:
```diff
@@ -0,0 +1,45 @@
+  box-shadow: var(--surface-shadow-large), var(--surface-shadow-large);
+  border: 1px solid var(--surface-color-border-default);
+}
+
```
Can we add a data-focus-visible like we do in button and link to make sure our focus outline is the same style as the rest of the site? That way we don't leave it up to the user to determine and get differing outlines?

My focus outline with green highlight color on MacOS
![Screenshot 2024-06-18 at 11 09 04 AM](https://github.com/bcgov/House-Innovations-HOUSNAV/assets/41478918/eec0cead-02a5-4d67-adf7-65404abb9145)

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,39 @@
+export interface WalkthroughCardProps extends ReactAriaLinkProps {
+  id: string;
+  data: WalkthroughInfo;
+  walkthroughId: string;
```
Can we add the data-testid prop like we have in the other components to match the paradigm? Lauren has said this is helpful for selection on her side for the cypress tests as well.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,39 @@
+    <ReactAriaLink
+      className="ui-WalkthroughCard--CardContainer"
+      href={`${URL_WALKTHROUGH_HREF}/${walkthroughId}`}
+      aria-label={`${id}-walkthrough-card`}
```
I like this as it doesn't read the whole card content aloud but I think we may need more context for the user. This is where I always struggle on determining enough context without too much. Something like:

`walkthrough ${id} - ${data.title}` ?

Would read out "walkthrough 9.9.9 [slight pause] Egress from Dwelling Units". Thoughts? Too much text/content?

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.tsx:
```diff
@@ -0,0 +1,39 @@
+      <article>
+        <header className="ui-WalkthroughCard--Title">{data.title}</header>
+        <span className="ui-WalkthroughCard--Subtitle">{data.subtitle}</span>
+        <section className="ui-WalkthroughCard--Description">
```
I think this can just be a `<p>` element. We don't need a separate section for the description IMO.

Also from MDN for `<section>` element:

"Sections should always have a heading, with very few exceptions."

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/section

nicholaslyon-dt commented on apps/web/app/page.tsx:
```diff
@@ -1,15 +1,19 @@
+      <div>
+        {Object.entries(WalkthroughJSONData).map(([id, { info }]) => {
+          return (
+            <WalkthroughCard key={id} id={id} data={info} walkthroughId={id} />
```
Also now since `id` and `walkthroughId` are the same I'm not sure we need separate props for that.

nicholaslyon-dt commented on packages/ui/src/walkthrough-card/WalkthroughCard.test.tsx:
```diff
@@ -0,0 +1,52 @@
+    expect(linkElement).toBeInTheDocument();
+  });
+
+  it("renders and navigates correctly on click", async () => {
```
Needs renamed to reflect it renders appropriately with the href, we don't test navigation anymore (which I agree with, don't think we need to test default functionality)

nicholaslyon-dt reviewed (changes requested):
Sorrrrrrrrrry I found more annoying things.

wardm89 committed:
remove data-focused css for walkthrough card

Walkthrough card refactor

nicholaslyon-dt reviewed (approved):
Love it. Looks fantastic!
